techlibs/xilinx/synth_xilinx.cc

```diff
@@ -441,6 +441,16 @@ struct SynthXilinxPass : public ScriptPass
 		}
 
 		if (check_label("map_ffram")) {
+			// Required for dffsr2dff to work.
+			run("simplemap t:$dff t:$adff t:$mux");
+			// Needs to be done before opt -mux_bool happens.
+			run("dffsr2dff");
```

Comment:
What's your take on making `dffsr2dff` be `$[a]dff` aware, just like `dff2dffe`?

Reply:
First, `dff2dffe` is not `$adff` aware — it only knows about `$dffe`.

Second, fixing that would require yet more cell types, and also making the `$__DFF*` cells into first-class primitives (so that we have something to simplemap into).  I'm not against it, but it's a *big* combinatorial explosion, and we'd need to discuss it properly.

Reply:
Sorry. I meant `$dff` (word level cells) aware. Long term, it would be good to preserve as much of the word level operations as possible, with a view to word level synthesis, but that's beyond the scope here.

Discussions with @cliffordwolf have led to the consensus we should add 16 new `$_DFFE_*` cells, since it is a common enough pattern.

Reply:
Which cells do you mean?  As far as I can tell, we're missing the following cells that could be plausibly advocated for here:

- word-level ff with async set/reset and enable
- word-level ff with sync set/reset
- word-level ff with sync set/reset and enable
- bit-level ff with async set/reset and enable (currently `$__DFFE_[NP][NP][01]` — 8 cells)
- bit-level ff with sync set/reset (currently `$__DFFS_[NP][NP][01]_` — 8 cells)
- bit-level ff with sync set/reset and enable (currently `$__DFFSE_[NP][NP][01]` — 8 cells)

This doesn't count further cell types that we would need to support FFs with both set and reset (another 3 coarse + 24 fine by my count)

Reply:
The other question is whether you then also add variants with enable-over-sync-reset priority; as required for iCE40 and I think Intel FPGAs; which would be another 8 cells

Reply:
Let's revisit this in a future PR